File: 695-MaxAreaOfIsland/695-MaxAreaOfIsland.cpp
// Last updated: 22/07/2025, 13:53:13
class Solution {
public:
    void dfs(vector<vector<int>>& grid, vector<vector<int>>& vis, int row, int col, int& area){
        vis[row][col] = 1;
        area += 1;
        // cout<<" area: "<< area<<endl;
        int drow[4] = {0,1,0,-1};
        int dcol[4] = {-1,0,1,0};
        int n = grid.size();
        int m = grid[0].size();
        for(int i = 0; i<4; i++){
            int nrow = drow[i] + row;
            int ncol = dcol[i] + col;
            if(nrow >= 0 && nrow <n && ncol >= 0 && ncol < m && vis[nrow][ncol] != 1 && grid[nrow][ncol] == 1){
                dfs(grid, vis, nrow, ncol,area);
            }
        }
    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int n = grid.size();
        int m  = grid[0].size();
        vector<vector<int>>vis(n, vector<int> (m, 0));
        int maxArea = 0;
        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                int area = 0;
                if(!vis[i][j] && grid[i][j] == 1){
                    dfs(grid, vis, i, j, area);
                    maxArea = max(maxArea, area);
                }
            }
        }
        return maxArea;
    }
};

File: 695-MaxAreaOfIsland/695-MaxAreaOfIsland.py
# Last updated: 01/10/2025, 00:12:42
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        
        def dfs(r,c):
            if r < 0 or r >= len(grid) or c < 0 or c >= len(grid[0]) or grid[r][c] == 0:
                return 0
            grid[r][c] = 0
            area = 1
            area += dfs(r+1, c)
            area += dfs(r-1,c)
            area += dfs(r, c+1)
            area += dfs(r, c - 1)
            return area
        max_area = 0
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == 1:
                    max_area = max(max_area, dfs(r, c))
        
        return max_area


        